File: javascript/designpattern/behavioral/Strategy.js
/**
 * Strategy Pattern Example
 * 
 * Intent: Define a family of algorithms, encapsulate each one, and make them interchangeable.
 * Strategy lets the algorithm vary independently from clients that use it.
 * 
 * Real-world analogy:
 * Think of a payment processing system where different payment methods
 * require different processing strategies:
 * 1. Credit Card processing with 3D Secure
 * 2. PayPal processing with OAuth
 * 3. UPI processing with QR code
 * 4. Bank Transfer processing with account verification
 * Each strategy encapsulates its own validation, processing, and security requirements.
 */

const { Payment, PaymentStatus } = require('./PaymentTheme');

// Strategy interface
class PaymentStrategy {
    validate(payment) {
        throw new Error('Method validate() must be implemented');
    }
    
    process(payment) {
        throw new Error('Method process() must be implemented');
    }
    
    refund(payment) {
        throw new Error('Method refund() must be implemented');
    }
}

// Concrete Strategies
class CreditCardStrategy extends PaymentStrategy {
    validate(payment) {
        console.log('\nValidating credit card payment...');
        
        // Validate card number (Luhn algorithm)
        if (!this.isValidCardNumber(payment.cardNumber)) {
            console.log('Invalid card number');
            return false;
        }
        
        // Validate expiry date
        if (!this.isValidExpiryDate(payment.expiryDate)) {
            console.log('Invalid expiry date');
            return false;
        }
        
        // Validate CVV
        if (!this.isValidCVV(payment.cvv)) {
            console.log('Invalid CVV');
            return false;
        }
        
        console.log('Credit card validation successful');
        return true;
    }
    
    process(payment) {
        console.log('\nProcessing credit card payment...');
        
        // Simulate 3D Secure verification
        if (this.requires3DSecure(payment)) {
            console.log('Initiating 3D Secure verification...');
            if (!this.verify3DSecure(payment)) {
                console.log('3D Secure verification failed');
                return false;
            }
            console.log('3D Secure verification successful');
        }
        
        // Simulate payment gateway processing
        console.log('Processing payment through payment gateway...');
        payment.status = PaymentStatus.COMPLETED;
        return true;
    }
    
    refund(payment) {
        console.log('\nProcessing credit card refund...');
        
        // Simulate refund through payment gateway
        console.log('Processing refund through payment gateway...');
        payment.status = PaymentStatus.REFUNDED;
        return true;
    }
    
    isValidCardNumber(cardNumber) {
        // Simple Luhn algorithm implementation
        const digits = cardNumber.split('').map(Number);
        const lastDigit = digits.pop();
        const sum = digits
            .reverse()
            .map((digit, index) => index % 2 === 0 ? digit * 2 : digit)
            .map(digit => digit > 9 ? digit - 9 : digit)
            .reduce((acc, digit) => acc + digit, 0);
        return (sum + lastDigit) % 10 === 0;
    }
    
    isValidExpiryDate(expiryDate) {
        const [month, year] = expiryDate.split('/');
        const currentDate = new Date();
        const currentYear = currentDate.getFullYear() % 100;
        const currentMonth = currentDate.getMonth() + 1;
        
        const expMonth = parseInt(month);
        const expYear = parseInt(year);
        
        return expYear > currentYear || 
               (expYear === currentYear && expMonth >= currentMonth);
    }
    
    isValidCVV(cvv) {
        return /^\d{3,4}$/.test(cvv);
    }
    
    requires3DSecure(payment) {
        return payment.amount > 1000;
    }
    
    verify3DSecure(payment) {
        // Simulate 3D Secure verification
        return Math.random() > 0.1; // 90% success rate
    }
}

class PayPalStrategy extends PaymentStrategy {
    validate(payment) {
        console.log('\nValidating PayPal payment...');
        
        // Validate email
        if (!this.isValidEmail(payment.email)) {
            console.log('Invalid PayPal email');
            return false;
        }
        
        // Simulate PayPal account verification
        if (!this.verifyPayPalAccount(payment)) {
            console.log('PayPal account verification failed');
            return false;
        }
        
        console.log('PayPal validation successful');
        return true;
    }
    
    process(payment) {
        console.log('\nProcessing PayPal payment...');
        
        // Simulate OAuth flow
        console.log('Initiating PayPal OAuth flow...');
        if (!this.authenticatePayPal(payment)) {
            console.log('PayPal authentication failed');
            return false;
        }
        
        // Simulate payment processing
        console.log('Processing payment through PayPal...');
        payment.status = PaymentStatus.COMPLETED;
        return true;
    }
    
    refund(payment) {
        console.log('\nProcessing PayPal refund...');
        
        // Simulate refund through PayPal
        console.log('Processing refund through PayPal...');
        payment.status = PaymentStatus.REFUNDED;
        return true;
    }
    
    isValidEmail(email) {
        return /^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(email);
    }
    
    verifyPayPalAccount(payment) {
        // Simulate PayPal account verification
        return Math.random() > 0.1; // 90% success rate
    }
    
    authenticatePayPal(payment) {
        // Simulate PayPal OAuth authentication
        return Math.random() > 0.1; // 90% success rate
    }
}

class UPIStrategy extends PaymentStrategy {
    validate(payment) {
        console.log('\nValidating UPI payment...');
        
        // Validate UPI ID
        if (!this.isValidUPIId(payment.upiId)) {
            console.log('Invalid UPI ID');
            return false;
        }
        
        // Validate QR code
        if (!this.validateQRCode(payment)) {
            console.log('Invalid QR code');
            return false;
        }
        
        console.log('UPI validation successful');
        return true;
    }
    
    process(payment) {
        console.log('\nProcessing UPI payment...');
        
        // Simulate UPI app authentication
        console.log('Initiating UPI app authentication...');
        if (!this.authenticateUPIApp(payment)) {
            console.log('UPI app authentication failed');
            return false;
        }
        
        // Simulate payment processing
        console.log('Processing payment through UPI...');
        payment.status = PaymentStatus.COMPLETED;
        return true;
    }
    
    refund(payment) {
        console.log('\nProcessing UPI refund...');
        
        // Simulate refund through UPI
        console.log('Processing refund through UPI...');
        payment.status = PaymentStatus.REFUNDED;
        return true;
    }
    
    isValidUPIId(upiId) {
        return /^[a-zA-Z0-9._-]+@[a-zA-Z]{3,}$/.test(upiId);
    }
    
    validateQRCode(payment) {
        // Simulate QR code validation
        return Math.random() > 0.1; // 90% success rate
    }
    
    authenticateUPIApp(payment) {
        // Simulate UPI app authentication
        return Math.random() > 0.1; // 90% success rate
    }
}

// Context
class PaymentProcessor {
    constructor(strategy) {
        this.strategy = strategy;
    }
    
    setStrategy(strategy) {
        this.strategy = strategy;
    }
    
    processPayment(payment) {
        console.log('\nProcessing payment with ' + this.strategy.constructor.name);
        
        if (this.strategy.validate(payment)) {
            if (this.strategy.process(payment)) {
                console.log('Payment processed successfully');
                return true;
            }
        }
        
        payment.status = PaymentStatus.FAILED;
        console.log('Payment processing failed');
        return false;
    }
    
    refundPayment(payment) {
        console.log('\nProcessing refund with ' + this.strategy.constructor.name);
        
        if (payment.status === PaymentStatus.COMPLETED) {
            if (this.strategy.refund(payment)) {
                console.log('Refund processed successfully');
                return true;
            }
        }
        
        console.log('Refund processing failed');
        return false;
    }
}

// Test the pattern
function testStrategy() {
    console.log('Strategy Pattern Demo - Payment Processing\n');
    
    // Create payment processor with credit card strategy
    const processor = new PaymentProcessor(new CreditCardStrategy());
    
    // Test credit card payment
    const creditCardPayment = new Payment(1500, 'USD', 'CREDIT_CARD');
    creditCardPayment.cardNumber = '4532015112830366';
    creditCardPayment.expiryDate = '12/25';
    creditCardPayment.cvv = '123';
    
    processor.processPayment(creditCardPayment);
    creditCardPayment.display();
    
    // Test refund
    processor.refundPayment(creditCardPayment);
    creditCardPayment.display();
    
    // Switch to PayPal strategy
    processor.setStrategy(new PayPalStrategy());
    
    // Test PayPal payment
    const paypalPayment = new Payment(500, 'USD', 'PAYPAL');
    paypalPayment.email = 'test@example.com';
    
    processor.processPayment(paypalPayment);
    paypalPayment.display();
    
    // Switch to UPI strategy
    processor.setStrategy(new UPIStrategy());
    
    // Test UPI payment
    const upiPayment = new Payment(200, 'USD', 'UPI');
    upiPayment.upiId = 'user@upi';
    
    processor.processPayment(upiPayment);
    upiPayment.display();
}

// Run the test
testStrategy(); 

File: java/designpattern/behavioral/Strategy.java
package designpattern.behavioral;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.util.regex.Pattern;

/**
 * Strategy Pattern Example
 * 
 * Intent: Define a family of algorithms, encapsulate each one, and make them interchangeable.
 * Strategy lets the algorithm vary independently from clients that use it.
 * 
 * Real-world analogy:
 * Think of a payment processing system where different payment methods
 * require different processing strategies:
 * 1. Credit Card processing with 3D Secure
 * 2. PayPal processing with OAuth
 * 3. UPI processing with QR code
 * 4. Bank Transfer processing with account verification
 * Each strategy encapsulates its own validation, processing, and security requirements.
 */

// Strategy interface
interface PaymentStrategy {
    boolean validate(Payment payment);
    boolean process(Payment payment);
    boolean refund(Payment payment);
}

// Concrete Strategies
class CreditCardStrategy implements PaymentStrategy {
    @Override
    public boolean validate(Payment payment) {
        System.out.println("\nValidating credit card payment...");
        
        // Validate card number (Luhn algorithm)
        if (!isValidCardNumber(payment.getCardNumber())) {
            System.out.println("Invalid card number");
            return false;
        }
        
        // Validate expiry date
        if (!isValidExpiryDate(payment.getExpiryDate())) {
            System.out.println("Invalid expiry date");
            return false;
        }
        
        // Validate CVV
        if (!isValidCVV(payment.getCvv())) {
            System.out.println("Invalid CVV");
            return false;
        }
        
        System.out.println("Credit card validation successful");
        return true;
    }
    
    @Override
    public boolean process(Payment payment) {
        System.out.println("\nProcessing credit card payment...");
        
        // Simulate 3D Secure verification
        if (requires3DSecure(payment)) {
            System.out.println("Initiating 3D Secure verification...");
            if (!verify3DSecure(payment)) {
                System.out.println("3D Secure verification failed");
                return false;
            }
            System.out.println("3D Secure verification successful");
        }
        
        // Simulate payment gateway processing
        System.out.println("Processing payment through payment gateway...");
        payment.setStatus(PaymentStatus.COMPLETED);
        return true;
    }
    
    @Override
    public boolean refund(Payment payment) {
        System.out.println("\nProcessing credit card refund...");
        
        // Simulate refund through payment gateway
        System.out.println("Processing refund through payment gateway...");
        payment.setStatus(PaymentStatus.REFUNDED);
        return true;
    }
    
    private boolean isValidCardNumber(String cardNumber) {
        // Simple Luhn algorithm implementation
        int[] digits = cardNumber.chars()
            .map(Character::getNumericValue)
            .toArray();
        
        int lastDigit = digits[digits.length - 1];
        int sum = 0;
        
        for (int i = digits.length - 2; i >= 0; i--) {
            int digit = digits[i];
            if ((digits.length - 1 - i) % 2 == 0) {
                digit *= 2;
                if (digit > 9) {
                    digit -= 9;
                }
            }
            sum += digit;
        }
        
        return (sum + lastDigit) % 10 == 0;
    }
    
    private boolean isValidExpiryDate(String expiryDate) {
        try {
            DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MM/yy");
            LocalDate expiry = LocalDate.parse("01/" + expiryDate, formatter);
            LocalDate current = LocalDate.now();
            return expiry.isAfter(current) || expiry.equals(current);
        } catch (Exception e) {
            return false;
        }
    }
    
    private boolean isValidCVV(String cvv) {
        return Pattern.matches("^\\d{3,4}$", cvv);
    }
    
    private boolean requires3DSecure(Payment payment) {
        return payment.getAmount() > 1000;
    }
    
    private boolean verify3DSecure(Payment payment) {
        // Simulate 3D Secure verification
        return Math.random() > 0.1; // 90% success rate
    }
}

class PayPalStrategy implements PaymentStrategy {
    @Override
    public boolean validate(Payment payment) {
        System.out.println("\nValidating PayPal payment...");
        
        // Validate email
        if (!isValidEmail(payment.getEmail())) {
            System.out.println("Invalid PayPal email");
            return false;
        }
        
        // Simulate PayPal account verification
        if (!verifyPayPalAccount(payment)) {
            System.out.println("PayPal account verification failed");
            return false;
        }
        
        System.out.println("PayPal validation successful");
        return true;
    }
    
    @Override
    public boolean process(Payment payment) {
        System.out.println("\nProcessing PayPal payment...");
        
        // Simulate OAuth flow
        System.out.println("Initiating PayPal OAuth flow...");
        if (!authenticatePayPal(payment)) {
            System.out.println("PayPal authentication failed");
            return false;
        }
        
        // Simulate payment processing
        System.out.println("Processing payment through PayPal...");
        payment.setStatus(PaymentStatus.COMPLETED);
        return true;
    }
    
    @Override
    public boolean refund(Payment payment) {
        System.out.println("\nProcessing PayPal refund...");
        
        // Simulate refund through PayPal
        System.out.println("Processing refund through PayPal...");
        payment.setStatus(PaymentStatus.REFUNDED);
        return true;
    }
    
    private boolean isValidEmail(String email) {
        return Pattern.matches("^[A-Za-z0-9+_.-]+@(.+)$", email);
    }
    
    private boolean verifyPayPalAccount(Payment payment) {
        // Simulate PayPal account verification
        return Math.random() > 0.1; // 90% success rate
    }
    
    private boolean authenticatePayPal(Payment payment) {
        // Simulate PayPal OAuth authentication
        return Math.random() > 0.1; // 90% success rate
    }
}

class UPIStrategy implements PaymentStrategy {
    @Override
    public boolean validate(Payment payment) {
        System.out.println("\nValidating UPI payment...");
        
        // Validate UPI ID
        if (!isValidUPIId(payment.getUpiId())) {
            System.out.println("Invalid UPI ID");
            return false;
        }
        
        // Validate QR code
        if (!validateQRCode(payment)) {
            System.out.println("Invalid QR code");
            return false;
        }
        
        System.out.println("UPI validation successful");
        return true;
    }
    
    @Override
    public boolean process(Payment payment) {
        System.out.println("\nProcessing UPI payment...");
        
        // Simulate UPI app authentication
        System.out.println("Initiating UPI app authentication...");
        if (!authenticateUPIApp(payment)) {
            System.out.println("UPI app authentication failed");
            return false;
        }
        
        // Simulate payment processing
        System.out.println("Processing payment through UPI...");
        payment.setStatus(PaymentStatus.COMPLETED);
        return true;
    }
    
    @Override
    public boolean refund(Payment payment) {
        System.out.println("\nProcessing UPI refund...");
        
        // Simulate refund through UPI
        System.out.println("Processing refund through UPI...");
        payment.setStatus(PaymentStatus.REFUNDED);
        return true;
    }
    
    private boolean isValidUPIId(String upiId) {
        return Pattern.matches("^[a-zA-Z0-9._-]+@[a-zA-Z]{3,}$", upiId);
    }
    
    private boolean validateQRCode(Payment payment) {
        // Simulate QR code validation
        return Math.random() > 0.1; // 90% success rate
    }
    
    private boolean authenticateUPIApp(Payment payment) {
        // Simulate UPI app authentication
        return Math.random() > 0.1; // 90% success rate
    }
}

// Context
class PaymentProcessor {
    private PaymentStrategy strategy;
    
    public PaymentProcessor(PaymentStrategy strategy) {
        this.strategy = strategy;
    }
    
    public void setStrategy(PaymentStrategy strategy) {
        this.strategy = strategy;
    }
    
    public boolean processPayment(Payment payment) {
        System.out.println("\nProcessing payment with " + strategy.getClass().getSimpleName());
        
        if (strategy.validate(payment)) {
            if (strategy.process(payment)) {
                System.out.println("Payment processed successfully");
                return true;
            }
        }
        
        payment.setStatus(PaymentStatus.FAILED);
        System.out.println("Payment processing failed");
        return false;
    }
    
    public boolean refundPayment(Payment payment) {
        System.out.println("\nProcessing refund with " + strategy.getClass().getSimpleName());
        
        if (payment.getStatus() == PaymentStatus.COMPLETED) {
            if (strategy.refund(payment)) {
                System.out.println("Refund processed successfully");
                return true;
            }
        }
        
        System.out.println("Refund processing failed");
        return false;
    }
}

// Test the pattern
public class Strategy {
    public static void main(String[] args) {
        System.out.println("Strategy Pattern Demo - Payment Processing\n");
        
        // Create payment processor with credit card strategy
        PaymentProcessor processor = new PaymentProcessor(new CreditCardStrategy());
        
        // Test credit card payment
        Payment creditCardPayment = new CreditCardPayment(1500.0, "USD", "4532015112830366", "12/25", "123");
        
        processor.processPayment(creditCardPayment);
        creditCardPayment.display();
        
        // Test refund
        processor.refundPayment(creditCardPayment);
        creditCardPayment.display();
        
        // Switch to PayPal strategy
        processor.setStrategy(new PayPalStrategy());
        
        // Test PayPal payment
        Payment paypalPayment = new PayPalPayment(500.0, "USD", "test@example.com");
        
        processor.processPayment(paypalPayment);
        paypalPayment.display();
        
        // Switch to UPI strategy
        processor.setStrategy(new UPIStrategy());
        
        // Test UPI payment
        Payment upiPayment = new UPIPayment(200.0, "USD", "user@upi");
        
        processor.processPayment(upiPayment);
        upiPayment.display();
    }
} 